File: src/services/interviewService.ts
import {
  InterviewSession,
  InterviewSetupPayload,
  SendMessagePayload,
  SendMessageResponse,
  ChatMessage
} from '../types/interview';
import { InterviewReport } from '../types/report';
import { aiGeneratorEngine } from './aiGeneratorEngine';
import { roadmapGeneratorService } from './roadmapGeneratorService';
import { MOCK_REPORTS } from './mockData';

const ACTIVE_SESSION_STORAGE_KEY = 'intervai_active_interview_session';
const REPORTS_ARCHIVE_STORAGE_KEY = 'intervai_reports_archive';

class InterviewService {
  private activeSession: InterviewSession | null = null;
  private reportsArchive: InterviewReport[] = [];

  constructor() {
    this.restoreSessionFromStorage();
    this.loadReportsArchiveFromStorage();
  }

  private loadReportsArchiveFromStorage() {
    if (typeof window === 'undefined') return;
    try {
      const raw = localStorage.getItem(REPORTS_ARCHIVE_STORAGE_KEY);
      if (raw) {
        this.reportsArchive = JSON.parse(raw);
      }
    } catch (err) {
      console.error('Failed to load reports archive from storage:', err);
    }
  }

  private saveReportsArchiveToStorage() {
    if (typeof window === 'undefined') return;
    try {
      localStorage.setItem(REPORTS_ARCHIVE_STORAGE_KEY, JSON.stringify(this.reportsArchive));
    } catch (err) {
      console.error('Failed to save reports archive to storage:', err);
    }
  }

  /**
   * Restores active interview session from localStorage on initialization
   */
  private restoreSessionFromStorage(): InterviewSession | null {
    if (typeof window === 'undefined') return null;
    try {
      const raw = localStorage.getItem(ACTIVE_SESSION_STORAGE_KEY);
      if (raw) {
        const parsed: InterviewSession = JSON.parse(raw);
        if (parsed && parsed.status === 'in_progress') {
          this.activeSession = parsed;
          return parsed;
        }
      }
    } catch (err) {
      console.error('Failed to restore active interview session from storage:', err);
    }
    return null;
  }

  /**
   * Persists active session to localStorage
   */
  public saveSessionToStorage(session?: InterviewSession): void {
    const sessionToSave = session || this.activeSession;
    if (!sessionToSave || typeof window === 'undefined') return;

    try {
      localStorage.setItem(ACTIVE_SESSION_STORAGE_KEY, JSON.stringify(sessionToSave));
    } catch (err) {
      console.error('Failed to save active session to localStorage:', err);
    }
  }

  /**
   * Clears active session from storage
   */
  public clearActiveSession(): void {
    this.activeSession = null;
    if (typeof window !== 'undefined') {
      try {
        localStorage.removeItem(ACTIVE_SESSION_STORAGE_KEY);
      } catch (err) {
        console.error('Failed to clear active session storage:', err);
      }
    }
  }

  async startInterview(payload: InterviewSetupPayload): Promise<InterviewSession> {
    // Check if an existing in_progress session is already active
    const restored = this.restoreSessionFromStorage();
    if (restored && restored.status === 'in_progress' && restored.topic === payload.topic) {
      return restored;
    }

    const initialQuestion = aiGeneratorEngine.generateInitialQuestion(
      payload.topic,
      payload.difficulty,
      payload.mode,
      payload.persona,
      payload.sourceDocuments,
      payload.targetRole,
      payload.candidateName
    );

    const session: InterviewSession = {
      id: `int-${Date.now()}`,
      candidateName: payload.candidateName || 'Alex Mercer',
      targetRole: payload.targetRole || 'Senior AI Engineer',
      topic: payload.topic,
      customTopics: payload.customTopics,
      difficulty: payload.difficulty,
      initialDifficulty: payload.difficulty,
      mode: payload.mode,
      persona: payload.persona,
      totalQuestions: payload.totalQuestions || 5,
      currentQuestionIndex: 1,
      startedAt: new Date().toISOString(),
      status: 'in_progress',
      messages: [initialQuestion],
      currentScore: 85,
      topicsCovered: [payload.topic],
      coveredConcepts: [],
      missingConcepts: [],
      durationSeconds: 0,
      sourceDocuments: payload.sourceDocuments,
      speechSettings: payload.speechSettings || {
        enabled: true,
        voiceURI: '',
        gender: 'Female',
        accent: 'US',
        rate: 1.0,
        pitch: 1.0,
        volume: 1.0,
        isMuted: false
      },
      difficultyHistory: [
        { questionNumber: 1, difficulty: payload.difficulty, score: 85 }
      ],
      enableProctoring: payload.enableProctoring ?? true
    };

    this.activeSession = session;
    this.saveSessionToStorage(session);
    return session;
  }

  async sendMessage(payload: SendMessagePayload): Promise<SendMessageResponse> {
    if (!this.activeSession || this.activeSession.id !== payload.interviewId) {
      const restored = this.restoreSessionFromStorage();
      if (!restored) {
        throw new Error('Interview session not found or expired.');
      }
    }

    const session = this.activeSession!;

    // Add candidate message to session transcript
    const userMsg: ChatMessage = {
      id: `msg-user-${Date.now()}`,
      sender: 'candidate',
      text: payload.message,
      timestamp: new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })
    };
    session.messages.push(userMsg);

    // Evaluate answer and generate next question dynamically
    const result = aiGeneratorEngine.evaluateAndGenerateNextQuestion(session, payload.message);

    // Attach evaluation to user message
    userMsg.evaluation = {
      score: result.score,
      feedback: result.feedback,
      followupSuggestion: result.followupSuggestion,
      coveredConcepts: result.coveredConcepts,
      missingConcepts: result.missingConcepts
    };

    // Update cumulative score & concepts covered in session
    session.currentScore = Math.round((session.currentScore + result.score) / 2);
    if (result.coveredConcepts.length > 0) {
      session.coveredConcepts = Array.from(new Set([...(session.coveredConcepts || []), ...result.coveredConcepts]));
    }
    if (result.missingConcepts.length > 0) {
      session.missingConcepts = Array.from(new Set([...(session.missingConcepts || []), ...result.missingConcepts]));
    }

    if (result.difficultyShifted && result.newDifficulty) {
      session.difficulty = result.newDifficulty;
    }

    let isComplete = false;

    if (result.nextQuestion) {
      session.currentQuestionIndex += 1;
      session.messages.push(result.nextQuestion);
      session.difficultyHistory.push({
        questionNumber: session.currentQuestionIndex,
        difficulty: session.difficulty,
        score: result.score
      });
    } else {
      session.status = 'completed';
      isComplete = true;
    }

    // Auto-Save session to storage after sending message
    this.saveSessionToStorage(session);

    return {
      evaluation: {
        score: result.score,
        feedback: result.feedback,
        followupSuggestion: result.followupSuggestion,
        coveredConcepts: result.coveredConcepts,
        missingConcepts: result.missingConcepts
      },
      nextQuestion: result.nextQuestion,
      isComplete,
      session,
      difficultyShifted: result.difficultyShifted,
      newDifficulty: result.newDifficulty
    };
  }

  getActiveSession(): InterviewSession | null {
    if (!this.activeSession) {
      this.restoreSessionFromStorage();
    }
    return this.activeSession;
  }

  /**
   * Complete & End Interview Session: generates final assessment report and stores it in archive
   */
  async endInterview(interviewId?: string): Promise<InterviewReport> {
    const session = this.activeSession || this.restoreSessionFromStorage();
    if (!session) {
      throw new Error('No active interview session found to complete.');
    }

    session.status = 'completed';
    const report = this.buildReportFromSession(session);

    // Save to memory and storage archives
    this.reportsArchive.unshift(report);
    this.saveReportsArchiveToStorage();

    // Clear active session
    this.clearActiveSession();

    return report;
  }

  async getReportsList(): Promise<InterviewReport[]> {
    this.loadReportsArchiveFromStorage();
    const reports = this.reportsArchive.length > 0 ? this.reportsArchive : MOCK_REPORTS;

    // Ensure all reports have generated roadmaps
    return reports.map(r => {
      if (!r.learningRoadmap) {
        r.learningRoadmap = roadmapGeneratorService.generateRoadmapForReport(r);
      } else {
        r.learningRoadmap = roadmapGeneratorService.restoreRoadmapProgress(r.learningRoadmap);
      }
      return r;
    });
  }

  async getReport(id?: string): Promise<InterviewReport> {
    this.loadReportsArchiveFromStorage();

    let targetReport: InterviewReport | undefined = undefined;

    if (id) {
      targetReport = this.reportsArchive.find(r => r.id === id || r.interviewId === id);
    }

    if (!targetReport && this.activeSession && (this.activeSession.id === id || !id)) {
      targetReport = this.buildReportFromSession(this.activeSession);
    }

    if (!targetReport) {
      targetReport = MOCK_REPORTS.find(r => r.id === id || r.interviewId === id);
    }

    if (!targetReport && this.reportsArchive.length > 0) {
      targetReport = this.reportsArchive[0];
    }

    if (!targetReport) {
      targetReport = this.buildReportFromSession({
        id: id || `int-${Date.now()}`,
        candidateName: 'Alex Mercer',
        targetRole: 'Senior AI Engineer',
        topic: 'Agentic AI',
        difficulty: 'Intermediate',
        initialDifficulty: 'Intermediate',
        mode: 'Technical',
        persona: 'Senior AI Engineer',
        totalQuestions: 5,
        currentQuestionIndex: 5,
        startedAt: new Date().toISOString(),
        status: 'completed',
        messages: [],
        currentScore: 88,
        topicsCovered: ['Agentic AI'],
        durationSeconds: 300,
        speechSettings: { enabled: true, voiceURI: '', gender: 'Female', accent: 'US', rate: 1, pitch: 1, volume: 1, isMuted: false },
        difficultyHistory: []
      });
    }

    // Attach or restore learning roadmap
    if (!targetReport.learningRoadmap) {
      targetReport.learningRoadmap = roadmapGeneratorService.generateRoadmapForReport(targetReport);
    } else {
      targetReport.learningRoadmap = roadmapGeneratorService.restoreRoadmapProgress(targetReport.learningRoadmap);
    }

    return targetReport;
  }

  private buildReportFromSession(session: InterviewSession): InterviewReport {
    const questions = session.messages
      .filter(m => m.sender === 'ai')
      .map((m, idx) => {
        const candidateMsg = session.messages[session.messages.indexOf(m) + 1];
        return {
          questionNumber: idx + 1,
          question: m.text,
          candidateAnswer: candidateMsg?.text || 'No response recorded',
          score: candidateMsg?.evaluation?.score || 85,
          aiFeedback: candidateMsg?.evaluation?.feedback || `Demonstrated understanding of ${session.topic}.`,
          sampleOptimalAnswer: `Optimal Senior ${session.topic} Answer addressing architecture, scale, and missing concepts (${(session.missingConcepts || []).join(', ') || 'N/A'}).`,
          topic: session.topic,
          difficultyLevel: m.difficultyLevel || session.difficulty
        };
      });

    const report: InterviewReport = {
      id: `rep-${session.id}`,
      interviewId: session.id,
      candidateName: session.candidateName,
      candidateEmail: 'alex.mercer@dev.io',
      date: new Date().toISOString().split('T')[0],
      durationMinutes: Math.max(1, Math.round(session.durationSeconds / 60)),
      topic: session.topic,
      difficulty: session.difficulty,
      initialDifficulty: session.initialDifficulty,
      mode: session.mode,
      persona: session.persona,
      personaVerdict: `Demonstrated technical depth in ${session.topic}. Covered concepts: ${(session.coveredConcepts || []).join(', ') || 'Core fundamentals'}. Recommended for Senior Engineer roles.`,
      overallScore: session.currentScore,
      hiringRecommendation: session.currentScore >= 88 ? 'Strong Hire' : session.currentScore >= 75 ? 'Hire' : 'Weak Hire',
      strengths: [
        `Clear articulation of ${session.topic} architecture mechanics.`,
        `Demonstrated good reasoning under ${session.persona} evaluation standards.`
      ],
      weaknesses: session.missingConcepts && session.missingConcepts.length > 0
        ? session.missingConcepts.map(c => `Could deepen technical explanation regarding ${c}.`)
        : ['Practice fine-tuning latency SLAs under distributed loads.'],
      recommendations: [
        `Review production trade-offs in ${session.topic}.`,
        'Practice system design edge-case handling.'
      ],
      topicScores: [
        { topic: session.topic, score: session.currentScore, fullMark: 100 }
      ],
      categoryBreakdown: [
        { category: 'Technical Accuracy', score: session.currentScore },
        { category: 'Concept Coverage', score: Math.min(95, session.currentScore + 2) },
        { category: 'Problem Solving', score: Math.max(70, session.currentScore - 2) },
        { category: 'Communication', score: 88 }
      ],
      difficultyTimeline: session.difficultyHistory,
      sourceDocuments: session.sourceDocuments,
      integritySummary: session.integritySummary,
      questions: questions.length > 0 ? questions : []
    };

    // Attach generated Learning Roadmap
    report.learningRoadmap = roadmapGeneratorService.generateRoadmapForReport(report);

    return report;
  }
}

export const interviewService = new InterviewService();

File: backend/services/interview_service.py
from datetime import datetime
from extensions import db
from models.interview import InterviewSession
from models.report import InterviewReport
from services.ai_question_generator import AIQuestionGeneratorService

class InterviewService:
    @staticmethod
    def get_active_session_for_user(user_id: str) -> InterviewSession | None:
        return InterviewSession.query.filter_by(user_id=user_id, status='in_progress').first()

    @staticmethod
    def get_session_by_id(session_id: str) -> InterviewSession | None:
        return InterviewSession.query.get(session_id)

    @staticmethod
    def start_interview(user_id: str, payload: dict) -> dict:
        # Check if user already has an active interview in progress
        existing_active = InterviewSession.query.filter_by(user_id=user_id, status='in_progress').first()
        if existing_active:
            return {
                'error': f"You already have an ongoing interview on '{existing_active.topic}'. Please complete or end it before starting a new one.",
                'session': existing_active.to_dict(),
                'status': 400
            }

        candidate_name = payload.get('candidateName', 'Candidate')
        target_role = payload.get('targetRole', 'Senior AI Engineer')
        topic = payload.get('topic', 'Agentic AI')
        difficulty = payload.get('difficulty', 'Intermediate')
        mode = payload.get('mode', 'Technical')
        persona = payload.get('persona', 'Senior AI Engineer')
        total_questions = payload.get('totalQuestions', 5)

        # Generate initial AI question
        initial_question = AIQuestionGeneratorService.generate_question(
            topic=topic,
            difficulty=difficulty,
            persona=persona,
            previous_questions=[]
        )

        session = InterviewSession(
            user_id=user_id,
            candidate_name=candidate_name,
            target_role=target_role,
            topic=topic,
            custom_topics=payload.get('customTopics', []),
            difficulty=difficulty,
            initial_difficulty=difficulty,
            mode=mode,
            persona=persona,
            total_questions=total_questions,
            current_question_index=1,
            status='in_progress',
            messages=[initial_question],
            topics_covered=[topic],
            source_documents=payload.get('sourceDocuments', []),
            speech_settings=payload.get('speechSettings', {}),
            difficulty_history=[{'questionNumber': 1, 'difficulty': difficulty, 'score': 0}],
            enable_proctoring=payload.get('enableProctoring', True)
        )

        db.session.add(session)
        db.session.commit()

        return {
            'message': 'Interview started successfully',
            'session': session.to_dict(),
            'status': 201
        }

    @staticmethod
    def save_session_progress(session_id: str, user_id: str, updates: dict) -> dict:
        session = InterviewSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return {'error': 'Interview session not found', 'status': 404}

        if 'durationSeconds' in updates:
            session.duration_seconds = updates['durationSeconds']
        if 'speechSettings' in updates:
            session.speech_settings = updates['speechSettings']
        if 'integrityLogs' in updates:
            session.integrity_logs = updates['integrityLogs']
        if 'currentQuestionIndex' in updates:
            session.current_question_index = updates['currentQuestionIndex']

        session.updated_at = datetime.utcnow()
        db.session.commit()

        return {'message': 'Session state auto-saved successfully', 'session': session.to_dict(), 'status': 200}

    @staticmethod
    def send_message(session_id: str, user_id: str, candidate_text: str, is_voice_input: bool = False) -> dict:
        session = InterviewSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return {'error': 'Session not found', 'status': 404}

        if session.status == 'completed':
            return {'error': 'This interview session has already been completed.', 'status': 400}

        # 1. Append candidate message
        cand_msg = {
            'id': f"msg-cand-{datetime.now().timestamp()}",
            'sender': 'candidate',
            'text': candidate_text,
            'timestamp': datetime.now().strftime('%I:%M %p')
        }

        # Find latest AI question text for evaluation
        previous_ai_messages = [m for m in session.messages if m['sender'] == 'ai']
        latest_q_text = previous_ai_messages[-1]['text'] if previous_ai_messages else session.topic

        # 2. Evaluate answer
        eval_result = AIQuestionGeneratorService.evaluate_answer(
            answer_text=candidate_text,
            question_text=latest_q_text,
            difficulty=session.difficulty
        )

        cand_msg['evaluation'] = eval_result

        # Update score running average
        prev_scores = [m['evaluation']['score'] for m in session.messages if m['sender'] == 'candidate' and 'evaluation' in m]
        all_scores = prev_scores + [eval_result['score']]
        avg_score = int(sum(all_scores) / len(all_scores))
        session.current_score = avg_score

        session.messages = session.messages + [cand_msg]

        # 3. Check if interview completed or generate next question
        is_complete = session.current_question_index >= session.total_questions

        next_q = None
        if not is_complete:
            session.current_question_index += 1

            # Generate follow-up or next unique question
            previous_questions = [m['text'] for m in session.messages if m['sender'] == 'ai']
            if len(session.messages) % 2 == 0:
                next_q = AIQuestionGeneratorService.generate_followup(candidate_text, session.topic, session.difficulty)
            else:
                next_q = AIQuestionGeneratorService.generate_question(session.topic, session.difficulty, session.persona, previous_questions)

            session.messages = session.messages + [next_q]
            session.difficulty_history = (session.difficulty_history or []) + [{
                'questionNumber': session.current_question_index,
                'difficulty': session.difficulty,
                'score': eval_result['score']
            }]

        db.session.commit()

        return {
            'evaluation': eval_result,
            'nextQuestion': next_q,
            'isComplete': is_complete,
            'session': session.to_dict(),
            'status': 200
        }

    @staticmethod
    def end_interview(session_id: str, user_id: str) -> dict:
        session = InterviewSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return {'error': 'Session not found', 'status': 404}

        session.status = 'completed'
        session.updated_at = datetime.utcnow()

        # Calculate final overall score
        cand_evals = [m['evaluation'] for m in session.messages if m['sender'] == 'candidate' and 'evaluation' in m]
        scores = [e['score'] for e in cand_evals]
        overall_score = int(sum(scores) / len(scores)) if scores else 82

        # Recommendation
        if overall_score >= 85:
            recommendation = 'Strong Hire'
        elif overall_score >= 75:
            recommendation = 'Hire'
        else:
            recommendation = 'Needs Improvement'

        # Generate Report
        date_str = datetime.now().strftime('%b %d, %Y')
        duration_mins = max(1, session.duration_seconds // 60)

        report = InterviewReport(
            interview_id=session.id,
            user_id=user_id,
            candidate_name=session.candidate_name,
            target_role=session.target_role,
            topic=session.topic,
            difficulty=session.difficulty,
            persona=session.persona,
            date=date_str,
            overall_score=overall_score,
            duration_minutes=duration_mins,
            hiring_recommendation=recommendation,
            persona_verdict=f"Candidate demonstrated solid technical competency in {session.topic} with clear problem-solving methodology.",
            category_scores=[
                {'category': 'Technical Knowledge', 'score': overall_score},
                {'category': 'Problem Solving', 'score': min(100, overall_score + 3)},
                {'category': 'Communication', 'score': max(60, overall_score - 4)},
                {'category': 'Architecture & Design', 'score': overall_score}
            ],
            question_evaluations=[
                {
                    'questionNumber': idx + 1,
                    'questionText': m.get('text', ''),
                    'candidateAnswer': session.messages[idx*2 + 1]['text'] if (idx*2 + 1) < len(session.messages) else '',
                    'score': cand_evals[idx]['score'] if idx < len(cand_evals) else 80,
                    'feedback': cand_evals[idx]['feedback'] if idx < len(cand_evals) else 'Good'
                }
                for idx, m in enumerate([msg for msg in session.messages if msg['sender'] == 'ai'][:session.total_questions])
            ],
            strengths=[f"Strong conceptual understanding of {session.topic}", "Clear structured articulation"],
            growth_areas=["Deep dive into edge-case latency SLAs", "Production deployment patterns"],
            integrity_summary={
                'score': 100,
                'riskLevel': 'Low',
                'metrics': {'tabSwitchCount': 0, 'focusLossCount': 0, 'pasteCount': 0}
            },
            learning_roadmap={
                'id': f"roadmap-{session.id}",
                'interviewTopic': session.topic,
                'overallProgressPercentage': 0,
                'estimatedCompletionTime': '14 Hours (7 Days)',
                'skillsToImprove': [
                    {
                        'id': 'skill-1',
                        'skillName': f"{session.topic} Advanced Patterns",
                        'priority': 'High',
                        'recommendedTopics': ['Concurrency', 'Distributed Locks'],
                        'practiceQuestions': [f"How do you benchmark {session.topic} bottlenecks?"],
                        'resources': [{'title': 'Official Documentation', 'url': 'https://docs.python.org', 'type': 'docs'}]
                    }
                ],
                'studyPlanTimeline': [
                    {'dayRange': 'Day 1-2', 'title': 'Theory Deep Dive', 'completed': False, 'tasks': ['Review architectural papers']}
                ]
            }
        )

        db.session.add(report)
        db.session.commit()

        return {
            'message': 'Interview completed successfully',
            'report': report.to_dict(),
            'status': 200
        }
